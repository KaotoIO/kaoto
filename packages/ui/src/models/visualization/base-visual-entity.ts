import { DefinedComponent } from '../camel-catalog-index';
import { BaseCamelEntity, EntityType } from '../camel/entities';
import { KaotoSchemaDefinition } from '../kaoto-schema';

/**
 * BaseVisualCamelEntity
 *
 * This interface is used to represent a visual Camel entity.
 * f.i. Camel Route, Kamelet, KameletBinding, etc.
 * All dedicated Camel code should be implemented using this interface.
 */
export interface BaseVisualCamelEntity extends BaseCamelEntity {
  id: string;
  type: EntityType;

  getId: () => string;

  setId: (id: string) => void;

  /** Given a path, get the component label */
  getNodeLabel: (path?: string, labelType?: string) => string;

  /** Given a path, get the component tooltip content */
  getTooltipContent: (path?: string) => string;

  /** Given a path, get the component type and definition */
  getComponentSchema: (path?: string) => VisualComponentSchema | undefined;

  /** Returnt fields that should be omitted when configuring this entity */
  getOmitFormFields: () => string[];

  /** Given a path, update the model */
  updateModel(path: string | undefined, value: unknown): void;

  /** Add a step to the underlying Camel entity */
  addStep: (options: {
    definedComponent: DefinedComponent;
    mode: AddStepMode;
    data: IVisualizationNodeData;
    targetProperty?: string;
  }) => void;

  /** Remove the step at a given path from the underlying Camel entity */
  removeStep: (path?: string) => void;

  /** Returns the NodeInteraction information so the UI can show whether this node can have children and/or siblings */
  getNodeInteraction(data: IVisualizationNodeData): NodeInteraction;

  /** Given a path, retrieve the Node validation status */
  getNodeValidationText(path?: string): string | undefined;

  /** Generates a IVisualizationNode from the underlying Camel entity */
  toVizNode: () => IVisualizationNode;
}

export interface BaseVisualCamelEntityConstructor {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (...args: any[]): BaseVisualCamelEntity;
  isApplicable: (entity: unknown) => boolean;
}

/**
 * IVisualizationNode
 *
 * This interface is used to represent a node in the visualization tree.
 * Usually, each IVisualizationNode represent a component in a BaseVisualCamelEntity.
 *
 * This interface shouldn't know anything about the specific BaseVisualCamelEntity.
 */
export interface IVisualizationNode<T extends IVisualizationNodeData = IVisualizationNodeData> {
  id: string;
  data: T;

  /** This property is only set on the root node */
  getBaseEntity(): BaseVisualCamelEntity | undefined;

  /** This method returns the label to be used by the canvas nodes */
  getNodeLabel(labelType?: string): string;

  /** This method returns the tooltip content to be used by the canvas nodes */
  getTooltipContent(): string;

  addBaseEntityStep(definedComponent: DefinedComponent, mode: AddStepMode, targetProperty?: string): void;

  getNodeInteraction(): NodeInteraction;

  getComponentSchema(): VisualComponentSchema | undefined;

  updateModel(value: unknown): void;

  getParentNode(): IVisualizationNode | undefined;

  setParentNode(parentNode?: IVisualizationNode): void;

  getPreviousNode(): IVisualizationNode | undefined;

  setPreviousNode(previousNode: IVisualizationNode): void;

  getNextNode(): IVisualizationNode | undefined;

  setNextNode(node: IVisualizationNode): void;

  getChildren(): IVisualizationNode[] | undefined;

  addChild(child: IVisualizationNode): void;

  removeChild(): void;

  populateLeafNodesIds(ids: string[]): void;

  /** Retrieve the node's validation status, relying into the underlying entity */
  getNodeValidationText(): string | undefined;
}

export interface IVisualizationNodeData {
  icon?: string;
  path?: string;
  entity?: BaseVisualCamelEntity;
  isPlaceholder?: boolean;
  isGroup?: boolean;
  [key: string]: unknown;
}

/**
 * VisualComponentSchema
 *
 * This interface is used to represent a component through
 * the name and the schema of the component.
 */
export interface VisualComponentSchema {
  title: string;
  schema: KaotoSchemaDefinition['schema'];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  definition: any;
}

export const enum AddStepMode {
  /** Used to append a new step before an existing step */
  PrependStep = 'prepend-step',
  /** Used to append a new step after an existing step */
  AppendStep = 'append-step',
  /** Used to replace an existing step */
  ReplaceStep = 'replace-step',
  /**
   * Used to insert a new step in the step property.
   * for instance, for adding steps to the `from` or `doTry` processor
   */
  InsertChildStep = 'insert-step',
  /**
   * Used to insert a new special step in a special step property.
   * for instance, for adding `when` clauses to the `choice.when` property,
   * or to add `doCatch` clauses to the `doTry.doCath` property
   */
  InsertSpecialChildStep = 'insert-special-step',
}

export interface NodeInteraction {
  canHavePreviousStep: boolean;
  canHaveNextStep: boolean;
  canHaveChildren: boolean;
  canHaveSpecialChildren: boolean;
  canReplaceStep: boolean;
  canRemoveStep: boolean;
  canRemoveFlow: boolean;
  canBeDisabled: boolean;
}
