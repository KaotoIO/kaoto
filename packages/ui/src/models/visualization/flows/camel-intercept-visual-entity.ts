import { Intercept, ProcessorDefinition } from '@kaoto/camel-catalog/types';
import { getCamelRandomId } from '../../../camel-utils/camel-random-id';
import { NodeIconResolver, NodeIconType, isDefined } from '../../../utils';
import { EntityType } from '../../camel/entities/base-entity';
import {
  BaseVisualCamelEntity,
  IVisualizationNode,
  IVisualizationNodeData,
  NodeInteraction,
} from '../base-visual-entity';
import { AbstractCamelVisualEntity } from './abstract-camel-visual-entity';
import { CamelComponentSchemaService } from './support/camel-component-schema.service';
import { CamelRouteVisualEntityData } from './support/camel-component-types';
import { CamelStepsService } from './support/camel-steps.service';
import { ModelValidationService } from './support/validators/model-validation.service';

export class CamelInterceptVisualEntity
  extends AbstractCamelVisualEntity<{ intercept: Intercept }>
  implements BaseVisualCamelEntity
{
  id: string;
  readonly type = EntityType.Intercept;
  static readonly ROOT_PATH = 'intercept';

  constructor(public interceptDef: { intercept: Intercept } = { intercept: {} }) {
    super(interceptDef);
    const id = interceptDef.intercept.id ?? getCamelRandomId(CamelInterceptVisualEntity.ROOT_PATH);
    this.id = id;
    this.interceptDef.intercept.id = id;
  }

  static isApplicable(interceptDef: unknown): interceptDef is { intercept: Intercept } {
    if (!isDefined(interceptDef) || Array.isArray(interceptDef) || typeof interceptDef !== 'object') {
      return false;
    }

    const objectKeys = Object.keys(interceptDef!);

    return objectKeys.length === 1 && this.ROOT_PATH in interceptDef! && typeof interceptDef.intercept === 'object';
  }

  getId(): string {
    return this.id;
  }

  setId(id: string): void {
    this.id = id;
    this.interceptDef.intercept.id = id;
  }

  getNodeInteraction(data: IVisualizationNodeData): NodeInteraction {
    const stepsProperties = CamelComponentSchemaService.getProcessorStepsProperties(
      (data as CamelRouteVisualEntityData).processorName as keyof ProcessorDefinition,
    );
    const canHavePreviousStep = CamelComponentSchemaService.canHavePreviousStep(
      (data as CamelRouteVisualEntityData).processorName,
    );
    const canHaveChildren = stepsProperties.find((property) => property.type === 'branch') !== undefined;
    const canHaveSpecialChildren = Object.keys(stepsProperties).length > 1;
    const canReplaceStep = data.path !== CamelInterceptVisualEntity.ROOT_PATH;
    const canRemoveStep = data.path !== CamelInterceptVisualEntity.ROOT_PATH;

    return {
      canHavePreviousStep,
      canHaveNextStep: canHavePreviousStep,
      canHaveChildren,
      canHaveSpecialChildren,
      canReplaceStep,
      canRemoveStep,
      canRemoveFlow: data.path === CamelInterceptVisualEntity.ROOT_PATH,
    };
  }

  getNodeValidationText(path?: string | undefined): string | undefined {
    const componentVisualSchema = this.getComponentSchema(path);
    if (!componentVisualSchema) return undefined;

    return ModelValidationService.validateNodeStatus(componentVisualSchema);
  }

  toVizNode(): IVisualizationNode<IVisualizationNodeData> {
    const interceptGroupNode = CamelStepsService.getVizNodeFromProcessor(
      CamelInterceptVisualEntity.ROOT_PATH,
      { processorName: CamelInterceptVisualEntity.ROOT_PATH as keyof ProcessorDefinition },
      this.interceptDef,
    );
    interceptGroupNode.data.entity = this;
    interceptGroupNode.data.isGroup = true;
    interceptGroupNode.data.icon = NodeIconResolver.getIcon(this.type, NodeIconType.VisualEntity);

    return interceptGroupNode;
  }

  toJSON(): { intercept: Intercept } {
    return { intercept: this.interceptDef.intercept };
  }

  protected getRootUri(): string | undefined {
    return undefined;
  }
}
