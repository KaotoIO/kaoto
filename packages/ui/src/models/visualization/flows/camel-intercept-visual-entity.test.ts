import { IVisualizationNodeData } from '../base-visual-entity';
import { CamelInterceptVisualEntity } from './camel-intercept-visual-entity';
import { ModelValidationService } from './support/validators/model-validation.service';

describe('CamelInterceptVisualEntity', () => {
  describe('constructor', () => {
    it('should allow to create an instance out of the object definition', () => {
      const interceptVisualEntity = new CamelInterceptVisualEntity({
        intercept: { id: 'a-reference', disabled: false },
      });

      expect(interceptVisualEntity.getId()).toEqual('a-reference');
    });

    it('should allow to create an instance out of the object definition without id', () => {
      const interceptRaw = {
        intercept: { id: undefined, uri: 'direct:a-reference' },
      };
      const interceptVisualEntity = new CamelInterceptVisualEntity(interceptRaw);

      expect(interceptVisualEntity.getId()).toBeDefined();
      expect(interceptRaw.intercept.id).toEqual(interceptVisualEntity.getId());
    });
  });

  describe('isApplicable', () => {
    it.each([
      [{ from: { id: 'from-1234', steps: [] } }, false],
      [{ onCompletion: { id: 'onCompletionId' } }, false],
      [{ onException: { id: 'onExceptionId' } }, false],
      [{ intercept: { id: 'interceptId' } }, true],
      [{ interceptFrom: { id: 'interceptFromId' } }, false],
      [{ interceptSendToEndpoint: { id: 'interceptSendToEndpointId' } }, false],
    ])('should return %s for %s', (definition, result) => {
      expect(CamelInterceptVisualEntity.isApplicable(definition)).toEqual(result);
    });
  });

  it('should return the id', () => {
    const interceptVisualEntity = new CamelInterceptVisualEntity({
      intercept: { id: 'id', disabled: false },
    });
    expect(interceptVisualEntity.getId()).toEqual('id');
  });

  it('should set the id', () => {
    const interceptVisualEntity = new CamelInterceptVisualEntity({ intercept: { id: 'a-reference', disabled: false } });
    interceptVisualEntity.setId('new-id');
    expect(interceptVisualEntity.getId()).toEqual('new-id');
    expect(interceptVisualEntity.interceptDef.intercept.id).toEqual('new-id');
  });

  describe('getNodeInteraction', () => {
    it.each([
      { processorName: 'route', path: 'route' },
      { processorName: 'from', path: 'from' },
      { processorName: 'to', path: 'to' },
      { processorName: 'log', path: 'log' },
      { processorName: 'onException', path: 'onException' },
      { processorName: 'onCompletion', path: 'onCompletion' },
      { processorName: 'intercept', path: 'intercept' },
      { processorName: 'interceptFrom', path: 'interceptFrom' },
      { processorName: 'interceptSendToEndpoint', path: 'interceptSendToEndpoint' },
    ] as const)(`should return the correct interaction for the '%s' processor`, (data) => {
      const interceptVisualEntity = new CamelInterceptVisualEntity({
        intercept: { id: 'id', disabled: false },
      });

      const result = interceptVisualEntity.getNodeInteraction(data as IVisualizationNodeData);
      expect(result).toMatchSnapshot();
    });
  });

  it('should delegate the validation text to the ModelValidationService', () => {
    const validateNodeStatusSpy = jest.spyOn(ModelValidationService, 'validateNodeStatus');

    const interceptVisualEntity = new CamelInterceptVisualEntity({
      intercept: { id: 'id', disabled: false },
    });
    interceptVisualEntity.getNodeValidationText('a-path');

    expect(validateNodeStatusSpy).toHaveBeenCalledWith(expect.anything());
  });

  it('should return the vizualization node', () => {
    const interceptVisualEntity = new CamelInterceptVisualEntity({
      intercept: { id: 'id', disabled: false },
    });
    const vizNode = interceptVisualEntity.toVizNode();

    expect(vizNode.data.processorName).toBe(CamelInterceptVisualEntity.ROOT_PATH);
    expect(vizNode.data.entity).toBe(interceptVisualEntity);
    expect(vizNode.data.isGroup).toBeTruthy();
  });

  it('should serialize the entity', () => {
    const interceptVisualEntity = new CamelInterceptVisualEntity({
      intercept: { id: undefined, disabled: false },
    });
    const result = interceptVisualEntity.toJSON();

    expect(result).toMatchSnapshot();
  });
});
