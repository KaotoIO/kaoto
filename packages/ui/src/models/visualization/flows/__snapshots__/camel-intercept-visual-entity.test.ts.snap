// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CamelInterceptVisualEntity getNodeInteraction should return the correct interaction for the '{
  processorName: 'interceptSendToEndpoint',
  path: 'interceptSendToEndpoint'
}' processor 1`] = `
{
  "canBeDisabled": false,
  "canHaveChildren": true,
  "canHaveNextStep": false,
  "canHavePreviousStep": false,
  "canHaveSpecialChildren": false,
  "canRemoveFlow": false,
  "canRemoveStep": true,
  "canReplaceStep": true,
}
`;

exports[`CamelInterceptVisualEntity getNodeInteraction should return the correct interaction for the '{ processorName: 'from', path: 'from' }' processor 1`] = `
{
  "canBeDisabled": false,
  "canHaveChildren": true,
  "canHaveNextStep": false,
  "canHavePreviousStep": false,
  "canHaveSpecialChildren": false,
  "canRemoveFlow": false,
  "canRemoveStep": true,
  "canReplaceStep": true,
}
`;

exports[`CamelInterceptVisualEntity getNodeInteraction should return the correct interaction for the '{ processorName: 'intercept', path: 'intercept' }' processor 1`] = `
{
  "canBeDisabled": false,
  "canHaveChildren": true,
  "canHaveNextStep": false,
  "canHavePreviousStep": false,
  "canHaveSpecialChildren": false,
  "canRemoveFlow": true,
  "canRemoveStep": false,
  "canReplaceStep": false,
}
`;

exports[`CamelInterceptVisualEntity getNodeInteraction should return the correct interaction for the '{ processorName: 'interceptFrom', path: 'interceptFrom' }' processor 1`] = `
{
  "canBeDisabled": false,
  "canHaveChildren": true,
  "canHaveNextStep": false,
  "canHavePreviousStep": false,
  "canHaveSpecialChildren": false,
  "canRemoveFlow": false,
  "canRemoveStep": true,
  "canReplaceStep": true,
}
`;

exports[`CamelInterceptVisualEntity getNodeInteraction should return the correct interaction for the '{ processorName: 'log', path: 'log' }' processor 1`] = `
{
  "canBeDisabled": false,
  "canHaveChildren": false,
  "canHaveNextStep": true,
  "canHavePreviousStep": true,
  "canHaveSpecialChildren": false,
  "canRemoveFlow": false,
  "canRemoveStep": true,
  "canReplaceStep": true,
}
`;

exports[`CamelInterceptVisualEntity getNodeInteraction should return the correct interaction for the '{ processorName: 'onCompletion', path: 'onCompletion' }' processor 1`] = `
{
  "canBeDisabled": false,
  "canHaveChildren": true,
  "canHaveNextStep": false,
  "canHavePreviousStep": false,
  "canHaveSpecialChildren": false,
  "canRemoveFlow": false,
  "canRemoveStep": true,
  "canReplaceStep": true,
}
`;

exports[`CamelInterceptVisualEntity getNodeInteraction should return the correct interaction for the '{ processorName: 'onException', path: 'onException' }' processor 1`] = `
{
  "canBeDisabled": false,
  "canHaveChildren": true,
  "canHaveNextStep": false,
  "canHavePreviousStep": false,
  "canHaveSpecialChildren": false,
  "canRemoveFlow": false,
  "canRemoveStep": true,
  "canReplaceStep": true,
}
`;

exports[`CamelInterceptVisualEntity getNodeInteraction should return the correct interaction for the '{ processorName: 'route', path: 'route' }' processor 1`] = `
{
  "canBeDisabled": false,
  "canHaveChildren": false,
  "canHaveNextStep": false,
  "canHavePreviousStep": false,
  "canHaveSpecialChildren": false,
  "canRemoveFlow": false,
  "canRemoveStep": true,
  "canReplaceStep": true,
}
`;

exports[`CamelInterceptVisualEntity getNodeInteraction should return the correct interaction for the '{ processorName: 'to', path: 'to' }' processor 1`] = `
{
  "canBeDisabled": false,
  "canHaveChildren": false,
  "canHaveNextStep": true,
  "canHavePreviousStep": true,
  "canHaveSpecialChildren": false,
  "canRemoveFlow": false,
  "canRemoveStep": true,
  "canReplaceStep": true,
}
`;

exports[`CamelInterceptVisualEntity should serialize the entity 1`] = `
{
  "intercept": {
    "disabled": false,
    "id": "intercept-1234",
  },
}
`;
